### PR TITLE
Editor: Improve behavior when right-clicking on multiple elements

### LINKED ISSUE #9426: Right Click Menu: Right Click from Element to Element Discovery
Currently the right click context menu opens up based on the selected element. This ticket is to:
1. Explore passing in an element to the callback handler to generate the menu items vs generating them based on the selected element
2. Explore alternatives to the fullscreen clickable background element approach to dismiss the context menu. The new approach should be geared towards allowing for right clicking between elements, but not cause any regressions from the previous approach.

Task 1 should be pretty straight forward to implement, however, task 2 could be a rabbit hole. Mainly setting this up as a 3 point Discovery ticket in case there's no fruitful approaches for task 2 (which is very much a possible outcome).

Please try to time box this. It's okay if nothing gets discovered. This is just a possible UX enhancement, so it's not worth spending a week on if initial discovery seems like there's no possible paths

### PR DESCRIPTION
**Pull request by maxyinger**
## Context
Part of investigative work to see if we could better the right click menu ux with a better pattern for dismissing it when clicking outside.

## Summary
Adds a new approach to detect clicking outside of the context menu to dismiss it.

## Relevant Technical Choices
Used the `mousedown` capture event on the document to always detect the `mousedown` event even if components have disabled bubbling or prevented default. This seemed to cover all needed mouse interaction scenarios to dismiss the context menu.

## To-do
NA

## User-facing changes
You should now be able to right click between elements in the canvas without needing to have an intermediary click to dismiss the context menu. Other than that, should behave the same as before

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Open the editor and make sure you have several elements of different types
2. right click one element and see the relevant right click menu appear
3. right click an element of another type and see the correct right click menu appear
4. click anywhere off the right click menu and see that the right click menu gets dismissed (this should behave the same as before)

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9426

### CODE REVIEW

**Review by BrittanyIRL — APPROVED** [review]
this is super cool, i like it a lot. Kinda weird that the dashboard karma tests are complaining, i'm out tomorrow so I'm approving this ahead of fixing that. great idea!

File: packages/design-system/src/utils/useComposeRefs.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Takes refs as a series of arguments and composes them to
+ * return a single callback ref.
+ *
+ * @param  {...*} refs - ref arguments
+ * @return {(node:HTMLElement) => void} composed callback ref
+ */
+function useComposeRefs(...refs) {
+  const refsRef = useRef(refs);
```
Comment (samwhale) [inline packages/design-system/src/utils/useComposeRefs.js:29]: refs ref lmao

File: packages/design-system/src/utils/useMouseDownOutsideRef.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * External dependencies
+ */
+import { useRef, useCallback } from '@web-stories-wp/react';
+/**
+ * Internal dependencies
+ */
+import { noop } from './noop';
+
+function useMouseDownOutsideRef(onPointerDownOutside = noop) {
```
Comment (samwhale) [inline packages/design-system/src/utils/useMouseDownOutsideRef.js]: 💭 Let's add a little docstring description here to describe what this does. Specifically that it returns a ref 😄

**Review by samwhale — APPROVED** [review]
Wow this is so much better. Cool thing is that right clicking another element now selects that element too since we don't have the mask. This is excellante
Screen Shot 2021-11-18 at 11 07 09 AM
Screen Shot 2021-11-18 at 11 07 22 AM

File: packages/design-system/src/utils/useMouseDownOutsideRef.js (hunk 2 above)
Comment (maxyinger) [reply packages/design-system/src/utils/useMouseDownOutsideRef.js]: ope. woops. thanks for pointing out

**Comment by maxyinger** [comment]
> Wow this is so much better. Cool thing is that right clicking another element now selects that element too since we don't have the mask. This is excellante Screen Shot 2021-11-18 at 11 07 09 AM Screen Shot 2021-11-18 at 11 07 22 AM

Not following what you're trying to point out with these screenshots. Mind elaborating?

**Comment by samwhale** [comment]
> > Wow this is so much better. Cool thing is that right clicking another element now selects that element too since we don't have the mask. This is excellante 
> 
> Not following what you're trying to point out with these screenshots. Mind elaborating?

hahaha shot wrong things. I mean to add my screen recordings 😮‍💨

**Comment by maxyinger** [comment]
> > > Wow this is so much better. Cool thing is that right clicking another element now selects that element too since we don't have the mask. This is excellante
> > 
> > 
> > Not following what you're trying to point out with these screenshots. Mind elaborating?
> 
> hahaha shot wrong things. I mean to add my screen recordings 😮‍💨  

lol right on

**Comment by maxyinger** [comment]
wow you're quick. Thanks for merging Pascal! 😄

**Pull request merged**